handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,471 @@
+        testServerHasNoStaple(SslProvider.OPENSSL);
+    }
+
+    @Test(timeout = 10000L)
+    public void testServerHasNoStapleOpenSslRefCnt() throws Exception {
+        testServerHasNoStaple(SslProvider.OPENSSL_REFCNT);
+    }
+
+    /**
+     * The server has OCSP stapling enabled but doesn't provide a staple.
+     */
+    private void testServerHasNoStaple(SslProvider sslProvider) throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        ChannelInboundHandlerAdapter serverHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                ctx.writeAndFlush(Unpooled.wrappedBuffer("Hello, World!".getBytes()));
+                ctx.fireChannelActive();
+            }
+        };
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                try {
+                    ReferenceCountUtil.release(msg);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        byte[] response = null;
+        TestClientOcspContext callback = new TestClientOcspContext(true);
+
+        handshake(sslProvider, latch, serverHandler, response, clientHandler, callback);
+
+        byte[] actual = callback.response();
+
+        assertNull(response);
+        assertNull(actual);
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientExceptionOpenSsl() throws Exception {
+        testClientException(SslProvider.OPENSSL);
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientExceptionOpenSslRefCnt() throws Exception {
+        testClientException(SslProvider.OPENSSL_REFCNT);
+    }
+
+    /**
+     * Testing what happens if the {@link ClientOcspContext} throws an {@link Exception}.
+     *
+     * The exception should bubble up on the client side and the connection should get closed.
+     */
+    private void testClientException(SslProvider sslProvider) throws Exception {
+        final AtomicReference<Throwable> causeRef = new AtomicReference<Throwable>();
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    causeRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        byte[] response = newOcspResponse();
+        OcspClientCallback callback = new OcspClientCallback() {
+            @Override
+            public boolean verify(byte[] response) throws Exception {
+                throw new OcspTestException("testClientException");
```

Comment:
consider creating this exception outside the scope and check for `assertSame` below:

```java
final OcspTestException clientException = new OcspTestException("testClientException");

...
@Override
public boolean verify(byte[] response) throws Exception {
  throw clientException;
}
...

assertSame(clientException, causeRef.get());
```